File: leetcode3/최원준/337. House Robber III.py
#

'''
1. 아이디어 :
bottom-up 방식으로 풀 수 있습니다.
dfs의 리턴값은 훔쳤을때의 최대값, 훔치지 않았을때의 최대값입니다.
- 현재 node를 훔치게 됐을때는, (현재node의 값 + 왼쪽에서 훔치지 않은 최대값 + 오른쪽에서 훔치지 않은 최대값),
- 현재 node를 훔치지 않을때는, max(왼쪽에서 훔쳤을때 최대값, 훔치지 않았을때 최대값) + max(오른쪽에서 훔쳤을때 최대값, 훔치지 않았을때 최대값)
마지막에 root에서는 둘 중 큰 값을 리턴합니다.

2. 시간복잡도 :
    O(n)

3. 자료구조/알고리즘 :
dfs

'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        def dfs(curr: TreeNode) -> list: # 훔쳤을때, 안훔쳤을떄
            if not curr:
                return [0, 0]
            curr_val = curr.val
            left_rob, left_clean = dfs(curr.left)
            right_rob, right_clean = dfs(curr.right)

            rob = left_clean + right_clean + curr_val
            clean = max(left_rob, left_clean) + max(right_rob, right_clean)

            return rob, clean

        rob, clean = dfs(root)
        return max(rob, clean)

File: leetcode3/이진희/337_House Robber III.java
/* 

1. 아이디어 : TreeNode와 DP의 성질을 활용
            후위순회로 리프 노드까지 이동 후, DP를 활용하여, 그 집을 털었을때, 털지 않았을때의 최댓값을 구해준다

2. 시간복잡도 : O(N)

3. 자료구조/알고리즘 : DP(DFS)

 */
class Solution {
    public int rob(TreeNode root) {
        // 루트부터 집을 털었을때 최댓값
        // 턴집과 연결되어 있는 집은 못 텀

        int[] ans = dfs(root);

        return Math.max(ans[0], ans[1]);

    }

    private int[] dfs(TreeNode node) {
        // 기저사례 
        if(node == null) return new int[]{0,0};

        int[] left = dfs(node.left);
        int[] right = dfs(node.right);

        int[] curr = new int[2];

        // 집을 털면 받을 수 있는 최댓값
        curr[0] = node.val + left[1] + right[1];

        //집을 안털면 받을 수 있는 최댓값
        curr[1] = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);

        return curr;
    }
}